File: app/src/main/java/dev/ragnarok/fenrir/view/DateTimePicker.java
package dev.ragnarok.fenrir.view;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Context;

import java.util.Calendar;
import java.util.Date;

import dev.ragnarok.fenrir.util.Logger;
import dev.ragnarok.fenrir.util.Unixtime;

public class DateTimePicker {

    private static final String TAG = DateTimePicker.class.getSimpleName();

    private final long time;
    private final Context context;
    private final Callback callback;

    private DateTimePicker(Builder builder) {
        time = builder.time;
        context = builder.context;
        callback = builder.callback;
    }

    private void show() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(time);

        int year = calendar.get(Calendar.YEAR);
        int month = calendar.get(Calendar.MONTH);
        int day = calendar.get(Calendar.DAY_OF_MONTH);
        int hours = calendar.get(Calendar.HOUR_OF_DAY);
        int minutes = calendar.get(Calendar.MINUTE);

        Logger.d(TAG, "onTimerClick, init time: " + new Date(time));

        new DatePickerDialog(context, (view, newYear, newMonth, newDay) ->
                showTime(newYear, newMonth, newDay, hours, minutes), year, month, day).show();
    }

    private void showTime(int year, int month, int day, int hour, int minutes) {
        new TimePickerDialog(context, (view, newHourOfDay, newMinutes) ->
                callback.onDateTimeSelected(Unixtime.of(year, month, day, newHourOfDay, newMinutes)), hour, minutes, true).show();
    }

    public interface Callback {
        void onDateTimeSelected(long unixtime);
    }

    public static class Builder {

        private final Context context;
        private Callback callback;
        private long time;

        public Builder(Context context) {
            this.context = context;
            time = System.currentTimeMillis();
        }

        public Builder setTime(long unixtime) {
            time = unixtime * 1000;
            return this;
        }

        public Builder setCallback(Callback callback) {
            this.callback = callback;
            return this;
        }

        public void show() {
            new DateTimePicker(this).show();
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/view/DateTimePicker.kt
package dev.ragnarok.fenrir.view

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Context
import android.widget.DatePicker
import android.widget.TimePicker
import dev.ragnarok.fenrir.util.Logger
import dev.ragnarok.fenrir.util.Unixtime
import java.util.Calendar
import java.util.Date

class DateTimePicker internal constructor(builder: Builder) {
    private val time: Long = builder.pTime
    private val context: Context = builder.context
    private val callback: Callback? = builder.pCallback
    internal fun show() {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = time
        val year = calendar[Calendar.YEAR]
        val month = calendar[Calendar.MONTH]
        val day = calendar[Calendar.DAY_OF_MONTH]
        val hours = calendar[Calendar.HOUR_OF_DAY]
        val minutes = calendar[Calendar.MINUTE]
        Logger.d(TAG, "onTimerClick, init time: " + Date(time))
        DatePickerDialog(
            context,
            { _: DatePicker?, newYear: Int, newMonth: Int, newDay: Int ->
                showTime(
                    newYear,
                    newMonth,
                    newDay,
                    hours,
                    minutes
                )
            },
            year,
            month,
            day
        ).show()
    }

    private fun showTime(year: Int, month: Int, day: Int, hour: Int, minutes: Int) {
        TimePickerDialog(
            context,
            { _: TimePicker?, newHourOfDay: Int, newMinutes: Int ->
                callback?.onDateTimeSelected(
                    Unixtime.of(
                        year,
                        month,
                        day,
                        newHourOfDay,
                        newMinutes
                    )
                )
            },
            hour,
            minutes,
            true
        ).show()
    }

    interface Callback {
        fun onDateTimeSelected(unixtime: Long)
    }

    class Builder(val context: Context) {
        var pCallback: Callback? = null
        var pTime: Long
        fun setTime(unixtime: Long): Builder {
            pTime = unixtime * 1000
            return this
        }

        fun setCallback(callback: Callback?): Builder {
            this.pCallback = callback
            return this
        }

        fun show() {
            DateTimePicker(this).show()
        }

        init {
            pTime = System.currentTimeMillis()
        }
    }

    companion object {
        private val TAG = DateTimePicker::class.java.simpleName
    }

}